### PR TITLE
Introducing support for Contexts (Config, Secret,Yaml, SecretYaml)

### PR DESCRIPTION
**Pull request by sandrogattuso**
Adding support to manage "Shared Configuration" for both resource and data source.
Supported contexts:

- config
- secret
- yaml
- secret-yaml

### CODE REVIEW

**Comment by logicbomb421** [comment]
Would love to see this released soon!

File: client/context.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	Type string                 `json:"type,omitempty"`
+	Data map[string]interface{} `json:"data,omitempty"`
+}
+
+func (context *Context) GetID() string {
+	return context.Metadata.Name
+}
+
+func (client *Client) GetContext(name string) (*Context, error) {
+	fullPath := fmt.Sprintf("/contexts/%s?decrypt=true", url.QueryEscape(name))
```
Comment (logicbomb421) [inline client/context.go:35]:
```suggestion
	fullPath := fmt.Sprintf("/contexts/%s?decrypt=true", url.PathEscape(name))
```
Using `QueryEscape` here produces odd results when `name` contains spaces. For example, `My Context/Other Piece' becomes `My+Context%2FOther+Piece`, which seems to return an HTTP 500 (confirmed via direct API call).

Using `PathEsacpe`, however, produces `My%20Context%2FOther%20Piece`, which _does_ return the context (and HTTP 200).

File: client/context.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+func (client *Client) UpdateContext(context *Context) (*Context, error) {
+
+	body, err := EncodeToJSON(context)
+
+	if err != nil {
+		return nil, err
+	}
+
+	fullPath := fmt.Sprintf("/contexts/%s", url.QueryEscape(context.Metadata.Name))
```
Comment (logicbomb421) [inline client/context.go:94]:
```suggestion
	fullPath := fmt.Sprintf("/contexts/%s", url.PathEscape(context.Metadata.Name))
```
Same reasoning here.

File: client/context.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		return nil, err
+	}
+
+	return &respContext, nil
+
+}
+
+func (client *Client) DeleteContext(name string) error {
+
+	fullPath := fmt.Sprintf("/contexts/%s", url.QueryEscape(name))
```
Comment (logicbomb421) [inline client/context.go:119]:
```suggestion
	fullPath := fmt.Sprintf("/contexts/%s", url.PathEscape(name))
```
Same reasoning here.

**Review by logicbomb421 — CHANGES_REQUESTED** [review]
Checked this out and ran it locally and noticed a few issues with context names that include spaces. Not entirely sure if this is an issue in the CF API, as `+` _is_ often recognized as encoding for a space, but does not seem to work here. The suggested fix handles spaces using standard url encoding, which does work.

**Comment by sandrogattuso** [comment]
> Checked this out and ran it locally and noticed a few issues with context names that include spaces. Not entirely sure if this is an issue in the CF API, as `+` _is_ often recognized as encoding for a space, but does not seem to work here. The suggested fix handles spaces using standard url encoding, which does work.

@logicbomb421 definitely not an issue with the API, not sure how I decided to use the QueryEscape function instead of the PathEscape. The value definitely belongs to the path, 100% agree with your feedback and the suggested changes.
I will push an update shortly and I'll also add a test case that contains special characters to validate this type of scenario.
Thanks for taking the time to provide the above input, really appreciated

**Comment by sandrogattuso** [comment]
@palson-cf also rebased this one, will do further rebase based on which PR is merged first

**Comment by sandrogattuso** [comment]
@palson-cf rebased again after last merge, this is now ready for review

**Review by palson-cf — APPROVED** [review]

**Pull request merged**